<a href="https://colab.research.google.com/github/Gnani0825/Gnaneshwar_INFO5731_Fall2024/blob/main/Pendyala_Gnaneshwar_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [ ]:
import math
#we are creating a space between r and v column
print("\tr\tV")
#now we created a function for the formula to return there result of maths expression
def volume(r):
  return (2/3)* math.pi * (r**3)
for r in range(2,21,2):
    hemi_vol=volume(r)
    print(f"\t{r}\t{hemi_vol:.2f}")

	r	V
	2	16.76
	4	134.04
	6	452.39
	8	1072.33
	10	2094.40
	12	3619.11
	14	5747.02
	16	8578.64
	18	12214.51
	20	16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [23]:
import random
num_ran =([random.randint(50, 150) for _ in range(12)])
print(" sequence of twelve random numbers:",num_ran)
#ascending list
ascen_nums=sorted(num_ran)
print(" numbers in ascending order:",ascen_nums)
#descending list
decen_nums=sorted(num_ran ,reverse=True)
print(" numbers in decending order:",decen_nums)
#average of random numbers
avg=sum(num_ran)/len(num_ran)
print(f" average of numbers:{avg:.2f}")
#average of first six ascending and descending order
avg_of_six_ascen_nums=sum(ascen_nums[:6])/6
avg_of_six_decen_nums=sum(decen_nums[:6])/6
print(f" average of first six numbers in ascending sorted list are: {avg_of_six_ascen_nums:.2f}")
print(f" average of first six numbers in decending sorted list are:{avg_of_six_decen_nums:.2f}")









 sequence of twelve random numbers: [95, 136, 80, 83, 88, 115, 137, 130, 97, 98, 101, 112]
 numbers in ascending order: [80, 83, 88, 95, 97, 98, 101, 112, 115, 130, 136, 137]
 numbers in decending order: [137, 136, 130, 115, 112, 101, 98, 97, 95, 88, 83, 80]
 average of numbers:106.00
 average of first six numbers in ascending sorted list are: 90.17
 average of first six numbers in decending sorted list are:121.83


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [36]:
import string
first=input("enter the first sentence: ")
second=input("enter the second sentence: ")
first_lis=first.split()
second_lis=second.split()
full_lis=first_lis + second_lis
print("\n transformed sentence is: ")
print(full_lis)
common=set(first_lis).intersection(set(second_lis))
print("\n list of words that occur in both sentences are: ")
print(common)
def remove_m(s):
  return s.translate(str.maketrans("", "", string.punctuation))
ganesh_one=remove_m(first)
ganesh_two=remove_m(second)
print("sanitized sentence one is: ",ganesh_one)
print("santized sentence two is: ", ganesh_two)
vowels="aeiou"
def vowels_c(r):
  vowel_count={vowel:0 for vowel in vowels}
  for char in r.lower():
    if char in vowel_count:
      vowel_count[char]+=1
  return vowel_count
vowel_one=vowels_c(ganesh_one)
vowel_two=vowels_c(ganesh_two)
print(f"sentence 1:",''.join([f"{v}:{c}" for v,c in vowel_one.items()]))
print(f"sentence 2:",''.join([f"{v}:{c}" for v,c in vowel_two.items()]))




enter the first sentence: hey gnaneshwar where are you .
enter the second sentence: my friend name is anil stays in irving

 transformed sentence is: 
['hey', 'gnaneshwar', 'where', 'are', 'you', '.', 'my', 'friend', 'name', 'is', 'anil', 'stays', 'in', 'irving']

 list of words that occur in both sentences are: 
set()
sanitized sentence one is:  hey gnaneshwar where are you 
santized sentence two is:  my friend name is anil stays in irving
sentence 1: a:3e:5i:0o:1u:1
sentence 2: a:3e:2i:6o:0u:0


# Question 4 (20 Points)
Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [28]:
import math
def calci():
  print("welcome to calculator")
  print("i can solve your problems")
  print("enter exit to quit the program")
  while True:
    math_expression=input("Enter the expression: ")
    if math_expression.lower() in ['exit','quit']:
      print("come again,good bye")
      break
  #we are using exception to catch the error to run program smoothly
    try:
       result = eval(math_expression, {"__builtins__": None}, {"math": math, "sqrt": math.sqrt})
       print(f"result: {result}")
    except ZeroDivisionError:
            print("Error: please do not ruin the application,division is not possible")
    except (SyntaxError, NameError):
            print("Error:  Please enter a valid  expression.")
    except Exception as e:
            print(f"Error: {e}")
calci()

welcome to calculator
i can solve your problems
enter exit to quit the program
Enter the expression: 2/3 * 4 *3
result: 8.0
Enter the expression: exit
come again,good bye


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [3]:
from google.colab import files


uploaded = files.upload()



Saving Assignment1_denton_housing.csv to Assignment1_denton_housing (2).csv


In [34]:
import pandas as pd
import io
from scipy import stats
#downloading the file
df = pd.read_csv(io.BytesIO(uploaded['Assignment1_denton_housing (2).csv']))
occupied_one = df[df['title_field'] == 'Occupied Housing Units']
occupied_one = occupied_one.sort_values(by='year')
#we now calculate yearly diff values and sort in ascending
occupied_one['yearly_diff'] = occupied_one['value'].diff()
yearly_diff = occupied_one.dropna(subset=['yearly_diff'])
sorted_diff = yearly_diff[['year', 'yearly_diff']].sort_values(by='yearly_diff')
print(sorted_diff)
pivot_df = df.pivot(index='year', columns='title_field', values='value')
pivot_df['Total Housing Units'] = pivot_df['Occupied Housing Units'] + pivot_df['Vacant Housing Units']
pivot_df['Vacant Percent'] = (pivot_df['Vacant Housing Units'] / pivot_df['Total Housing Units']) * 100
pivot_df['Occupied Percent'] = (pivot_df['Occupied Housing Units'] / pivot_df['Total Housing Units']) * 100
#we now calculate housing percent
for year in pivot_df.index:
     print(f"{year}  {pivot_df.loc[year, 'Vacant Percent']:.2f}%   {pivot_df.loc[year, 'Occupied Percent']:.2f}%")
maximum_vacant_one = pivot_df['Vacant Housing Units'].idxmax(), pivot_df['Vacant Housing Units'].max()
maximum_occupied_one = pivot_df['Occupied Housing Units'].idxmax(), pivot_df['Occupied Housing Units'].max()
#cwe calculate highest number of units of vaccant and occupied houses
print("\nYear    Value")
print(f"Vacant Housing Units   {maximum_vacant_one[0]}   {maximum_vacant_one[1]}")
print(f"Occupied Housing Units {maximum_occupied_one[0]} {maximum_occupied_one[1]}")
total = pivot_df['Total Housing Units']
harmonic_mean = stats.hmean(total)
print(f"\nharmonic Mean of total housing units: {harmonic_mean:.2f}")




    year  yearly_diff
17  2009       -795.0
11  2011       -303.0
8   2012       1404.0
5   2013       1565.0
2   2014       2622.0
14  2010       5091.0
2008  10.88%   89.12%
2009  9.07%   90.93%
2010  8.14%   91.86%
2011  9.38%   90.62%
2012  6.68%   93.32%
2013  8.80%   91.20%
2014  5.73%   94.27%

Year    Value
Vacant Housing Units   2008   4483
Occupied Housing Units 2014 46295

Harmonic Mean of Total Housing Units: 44398.29


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# write your responce below
'''
As i  already told i have a little idea about java and this format is completlely different from that,the assignment was challenging like i have no idea about the last question it made me to do some reasearch about the concept and finnaly done with that,and i feel the energy when i solved the problem and the first three are simple one's.yes,the time for the assignment is ok,but i know the basic part and few problems are beyond that level,i improved a lot when i compare to last week and i need to concentrate on the logic problems to get the grip in the coding.
'''
#thankyou
